crypto/ml_kem/ml_kem.c

```diff
@@ -0,0 +1,1567 @@
+
+    for (i = 0; i < DEGREE / 2; i++) {
+        real_real = (uint32_t)lhs->c[2 * i] * rhs->c[2 * i];
+        img_img = (uint32_t)lhs->c[2 * i + 1] * rhs->c[2 * i + 1];
+        real_img = (uint32_t)lhs->c[2 * i] * rhs->c[2 * i + 1];
+        img_real = (uint32_t)lhs->c[2 * i + 1] * rhs->c[2 * i];
+        out->c[2 * i] =
+            reduce(real_real +
+                   (uint32_t)reduce(img_img) * kModRoots[i]);
+        out->c[2 * i + 1] = reduce(img_real + real_img);
+    }
+}
+
+static ossl_inline
+void scalar_mult_add(scalar *out, const scalar *lhs,
+                     const scalar *rhs)
+{
+    scalar product;
+
+    scalar_mult(&product, lhs, rhs);
+    scalar_add(out, &product);
+}
+
+static const uint8_t kMasks[8] = {0x01, 0x03, 0x07, 0x0f,
+                                  0x1f, 0x3f, 0x7f, 0xff};
+
+/*
+ * FIPS 203, Section 4.2.1, Algorithm 5: ByteEncode_d, for 2<=d<=12 Here |bits|
+ * is |d|.  For efficiency, we handle the d=1 case separately.
+ */
+static void scalar_encode(uint8_t *out, const scalar *s, int bits)
+{
+    uint8_t out_byte = 0;
+    int out_byte_bits = 0;
+    int i, element_bits_done, chunk_bits, out_bits_remaining;
+    uint16_t element;
+
+    assert(bits <= (int)sizeof(*s->c) * 8 && bits != 1);
+    for (i = 0; i < DEGREE; i++) {
+        element = s->c[i];
+        element_bits_done = 0;
+        while (element_bits_done < bits) {
+            chunk_bits = bits - element_bits_done;
+            out_bits_remaining = 8 - out_byte_bits;
+            if (chunk_bits >= out_bits_remaining) {
+                chunk_bits = out_bits_remaining;
+                out_byte |= (element & kMasks[chunk_bits - 1]) << out_byte_bits;
+                *out = out_byte;
+                out++;
+                out_byte_bits = 0;
+                out_byte = 0;
+            } else {
+                out_byte |= (element & kMasks[chunk_bits - 1]) << out_byte_bits;
+                out_byte_bits += chunk_bits;
+            }
+            element_bits_done += chunk_bits;
+            element >>= chunk_bits;
+        }
+    }
+    if (out_byte_bits > 0)
+        *out = out_byte;
+}
+
+/*
+ * scalar_encode_12 is |scalar_encode| specialised for |bits| == 12.
+ */
+static void scalar_encode_12(uint8_t *out, const scalar *s)
+{
+    const uint16_t *c = s->c;
+    int i;
+
+    for (i = 0; i < DEGREE / 2; ++i) {
+        uint16_t c1 = *c++;
+        uint16_t c2 = *c++;
+
+        *out++ = (uint8_t) c1;
+        *out++ = (uint8_t) (((c1 >> 8) & 0x0f) | ((c2 & 0x0f) << 4));
+        *out++ = (uint8_t) (c2 >> 4);
+    }
+}
+
+/*
+ * scalar_encode_1 is |scalar_encode| specialised for |bits| == 1.
+ * uint8_t out[32]
+ */
+static void scalar_encode_1(uint8_t *out, const scalar *s)
+{
+    int i, j;
+    uint8_t out_byte;
+
+    for (i = 0; i < DEGREE; i += 8) {
+        out_byte = 0;
+        for (j = 0; j < 8; j++)
+            out_byte |= (s->c[i + j] & 1) << j;
+        *out = out_byte;
+        out++;
```

Comment:
not sure why this isnt just done on the line before.

Reply:
Boring SSL code I had not touched. Done (shortly) in #26172.